text/3617-precise-capturing.md

```diff
@@ -0,0 +1,668 @@
+//                 ^^^^^^^^^^^^^^^^
+//                 ^ Captures nothing.
+```
+
+Similarly, we can use this to avoid overcapturing a lifetime parameter so as to migrate code to Rust 2024:;
+
+```rust
+fn foo<'t>(_: &'t ()) -> impl use<> Sized {}
+//                       ^^^^^^^^^^^^^^^^
+//                       ^ Captures nothing.
+```
+
+We can use this to capture some generic parameters but not others:
+
+```rust
+fn foo<'t, T, U>(_: &'t (), _: T, y: U) -> impl use<U> Sized { y }
+//                                         ^^^^^^^^^^^^^^^^^
+//                                         ^ Captures `U` only.
+```
+
+## Generic const parameters
+
+In addition to type and lifetime parameters, we can use this to capture generic const parameters:
+
+```rust
+fn foo<'t, const C: u8>(_: &'t ()) -> impl use<C> Sized { C }
+//                                    ^^^^^^^^^^^^^^^^^
+//                                    ^ Captures `C` only.
+```
+
+## Capturing from outer inherent impl
+
+We can capture generic parameters from an outer inherent impl:
+
+```rust
+struct Ty<'a, 'b>(&'a (), &'b ());
+
+impl<'a, 'b> Ty<'a, 'b> {
+    fn foo(x: &'a (), _: &'b ()) -> impl use<'a> Sized { x }
+    //                              ^^^^^^^^^^^^^^^^^^
+    //                              ^ Captures `'a` only.
+}
+```
+
+## Capturing from outer trait impl
+
+We can capture generic parameters from an outer trait impl:
+
+```rust
+trait Trait<'a, 'b> {
+    type Foo;
+    fn foo(_: &'a (), _: &'b ()) -> Self::Foo;
+}
+
+impl<'a, 'b> Trait<'a, 'b> for () {
+    type Foo = impl use<'a> Sized;
+    //         ^^^^^^^^^^^^^^^^^^
+    //         ^ Captures `'a` only.
+    fn foo(x: &'a (), _: &'b ()) -> Self::Foo { x }
+}
+```
+
+## Capturing in trait definition
+
+We can capture generic parameters from the trait definition:
+
+```rust
+trait Trait<'a, 'b> {
+    fn foo(_: &'a (), _: &'b ()) -> impl use<'a, Self> Sized;
+    //                              ^^^^^^^^^^^^^^^^^^^^^^^^
+    //                              ^ Captures `'a` and `Self` only.
+}
+```
+
+## Capturing elided lifetimes
+
+We can capture elided lifetimes:
+
+```rust
+fn foo(x: &()) -> impl use<'_> Sized { x }
+//                ^^^^^^^^^^^^^^^^^^
+//                ^ Captures `'_` only.
+```
+
+## Combining with `for<..>`
+
+The `use<..>` specifier applies to the entire `impl Trait` opaque type.  In contrast, a `for<..>` binder applies to an individual *bound* within an opaque type.  Therefore, when both are used within the same type, `use<..>` always appears first.  E.g.:
+
+```rust
+fn foo<T>(_: T) -> impl use<T> for<'a> FnOnce(&'a ()) { |&()| () }
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Syntax
+
+The [syntax for `impl Trait`][] is revised and extended as follows:
+
+[syntax for `impl Trait`]: https://doc.rust-lang.org/nightly/reference/types/impl-trait.html
+
+> _UseCaptures_ :\
+> &nbsp;&nbsp; `use` [_GenericParams_][]
```

Comment:
*GenericParams* implies `impl use<T: Copy, 'a: 'static, const N: usize> Trait` is a valid parse,   
while `impl use<..> Trait` and `impl use<> Trait` are not.

Better create a separate rule.

```ebnf
UseCaptureParams ::=
    '<'  '..'  '>'
|   '<'  (UseCaptureParam ',')* UseCaptureParam? '>'

UseCaptureParam ::=
    LIFETIME_OR_LABEL
|   IDENTIFIER
```